docs/RFD-JobWorkerService.md

```diff
@@ -0,0 +1,366 @@
+  
+  - _`MemBytes`_ - maximum amount of memory used by the job. 
+  This is configured by setting [memory.max](https://docs.kernel.org/admin-guide/cgroup-v2.html). 
+  
+  - _`IOBytesPerSecond`_ - maximum read and write on the device mounted / is mounted on. 
+  This is configured by setting [io.max](https://docs.kernel.org/admin-guide/cgroup-v2.html) in the cgroup for the process. 
+  For example, a IOBytesPerSecond of 1000000000 would be similar to `259:1 rbps=1000000000 wbps=1000000000 riops=max wiops=max` in the cgroup's `io.max` file. 
+
+  - _`Command`_ - command to execute. For example, `_/bin/bash_`. 
+  
+  - _`Arguments`_ - arguments to pass to the command. 
+  For example, `"-c" "echo hello"` would be provided to command and ultimately run similar to `/bin/bash -c "echo hello".`
+
+
+* **func New(config JobConfig) \*Job** - create new Job instance, based on provided job configuration.
+
+
+* **func (\*Job) Start() error** - start job executions and return [error](https://pkg.go.dev/errors) if job failed to start or required fields contains no value.
+
+  Starting new job tart takes following steps:
+
+  - create new [cgroup](https://docs.kernel.org/admin-guide/cgroup-v2.html) for the job - e.g `/sys/fs/cgroup/<UUID>`
+  - create subtree to control resources - e.g. `+cpu +memory +io` into file `/sys/fs/cgroup/<UUID>/cgroup.subtree_control`
+  - set value for CPU, Memory and IO limits based on job configuration 
+    - `CPU` - set by writing QUOTA/PERIOD into `/sys/fs/cgroup/<UUID>/tasks/cpu.max`
+    - `Memory` - set by writing the provided IOBytesPerSecond to `/sys/fs/cgroup/<uuid>/tasks/memory.max`
+    - `IO` - is set by writing IOBytesPerSecond into `<MAJOR NUMBER OF PHYSICAL DEVICE>:<MINOR NUMBER OF PHYSICAL DEVICE> rbps=<IOBytesPerSecond> wbps=<IOBytesPerSecond> riops=max wiops=max` to `/sys/fs/cgroup/<UUID>/tasks/io.max` <br/><br/> _Note_:  `<MAJOR NUMBER OF PHYSICAL DEVICE>` and `<MINOR NUMBER OF PHYSICAL DEVICE>` shoudl be dynamically detected and provided during runtime.
+  - configures a new [exec.Cmd](https://pkg.go.dev/os/exec#Cmd)
+    - sets clone flags for creating new mount, pid, and network namespaces 
+    - sets [unshare flags](https://pkg.go.dev/syscall#SysProcAttr) for mount so that new mounts are not reflected in host mount namespace 
+    - sets [UseCgroupFD and CgroupFD](https://pkg.go.dev/syscall#SysProcAttr) to the cgroup's file descriptor such as `/sys/fs/cgroup/<UUID>/tasks`
+  - runs the configured `exec.Cmd` in a new goroutine
+    - does not wait for the command to complete 
+    - the goroutine will stop when the process completes or stopped
+
+
+* **func (\*Job) Status() JobStatus** - returns the current status of the job (see [JobStatus](#type-JobStatus))
+
+
+* **func (\*Job) Stream() OutputReader** - returns an `OutputReader` (type that implements [io.Reader](https://pkg.go.dev/io#Reader)). 
```

Comment:
How a client will be able to cancel ongoing stream ?

Reply:
@smallinsky  good point :)) 

I've added following section in RTD:  https://github.com/P-A-R-U-S/Go-Job-Worker-Service/pull/1/commits/7313c5d0bd1527e591921de17d85fdbbb698ae83
![Screenshot 2024-08-26 at 10 00 03 PM](https://github.com/user-attachments/assets/404dea6c-3245-4a52-823c-2f7beb432949)

But I don't see value to "overload" design document with actual implementation. It is draft code for now, but in general implementation would looks like this:
![Screenshot 2024-08-26 at 9 36 39 PM](https://github.com/user-attachments/assets/37608d9c-543b-492e-9725-27cb392822c7)

I didn't do PR, because I need to approve RFD first, but here is branch: https://github.com/P-A-R-U-S/Go-Job-Worker-Service/tree/grpc-server-and-client (as I mention all code are not ready for final review)

Reply:
I didn't put this into document, because it is more to implementation part, but approach would looks like:

```
Create a new background context with WithCancel, and then you can call cancel() on `defer`.

ctx, cancel := context.WithCancel(context.Background())
stream, err := client.StreamingRPC(ctx)
// do something on stream
cancel()
```
To gracefully handle this - client just need to close client. I

On the server side we need to use [signal.Notify](https://pkg.go.dev/os/signal#example-Notify)  to call [Server.GracefulStop](https://pkg.go.dev/google.golang.org/grpc#Server.GracefulStop)

Reply:
This will work on grpc client->server side, but what about server-side ->library usage?

Right now, the Job interface is defined as:
```
func New(config JobConfig) *Job 
func (*Job) Stream() OutputReader
```

Currently, there is no way for the library user to cancel the `Stream()` call.

What would you think about returning an `io.ReadCloser` object from `func (*Job) Stream() io.ReadCloser` and allow job `Stream` consumer to cancel the streaming in any time ?

Reply:
@smallinsky agree, make sense. Updated.
![Screenshot 2024-08-29 at 12 32 42 PM](https://github.com/user-attachments/assets/ebf21e57-1d78-4d40-9a21-5cd787ff2f19)